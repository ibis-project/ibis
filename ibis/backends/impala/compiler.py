from __future__ import annotations

import ibis.expr.operations as ops
from ibis.backends.base.sql.compiler import Compiler, ExprTranslator, TableSetFormatter
from ibis.backends.base.sql.registry import binary_infix_ops, operation_registry, unary


class ImpalaTableSetFormatter(TableSetFormatter):
    def _get_join_type(self, op):
        jname = self._join_names[type(op)]

        # Impala requires this
        if not op.predicates:
            jname = self._join_names[ops.CrossJoin]

        return jname

    def _format_in_memory_table(self, op):
        if op.data:
            return super()._format_in_memory_table(op)

        schema = op.schema
        names = schema.names
        types = schema.types
        rows = [
            f"{self._translate(ops.Cast(ops.Literal(None, dtype=dtype), to=dtype))} AS {name}"
            for name, dtype in zip(map(self._quote_identifier, names), types)
        ]
        return f"(SELECT * FROM (SELECT {', '.join(rows)}) AS _ LIMIT 0)"


class ImpalaExprTranslator(ExprTranslator):
    _registry = {**operation_registry, **binary_infix_ops, ops.Hash: unary("fnv_hash")}
    _forbids_frame_clause = (
        *ExprTranslator._forbids_frame_clause,
        ops.Lag,
        ops.Lead,
        ops.FirstValue,
        ops.LastValue,
    )
    _unsupported_reductions = (
        ops.ApproxMedian,
        ops.ApproxCountDistinct,
        ops.GroupConcat,
    )
    _dialect_name = "hive"
    _quote_identifiers = True


rewrites = ImpalaExprTranslator.rewrites


@rewrites(ops.FloorDivide)
def _floor_divide(op):
    return ops.Floor(ops.Divide(op.left, op.right))


class ImpalaCompiler(Compiler):
    translator_class = ImpalaExprTranslator
    table_set_formatter_class = ImpalaTableSetFormatter
