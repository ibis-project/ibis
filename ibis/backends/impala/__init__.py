"""Impala backend"""
import ibis.config
from ibis.backends.base import BaseBackend

# these objects are exposed in the public API and are not used in the module
from .client import (  # noqa: F401
    ImpalaClient,
    ImpalaConnection,
    ImpalaDatabase,
    ImpalaDatabaseTable,
    ImpalaTable,
)
from .hdfs import HDFS, WebHDFS, hdfs_connect
from .udf import *  # noqa: F401,F403


class Backend(BaseBackend):
    name = 'impala'
    kind = 'sql'
    client = ImpalaClient
    database_class = ImpalaDatabase
    table_class = ImpalaDatabaseTable
    table_expr_class = ImpalaTable
    HDFS = HDFS
    WebHDFS = WebHDFS

    def hdfs_connect(self, *args, **kwargs):
        return hdfs_connect(*args, **kwargs)

    def connect(
        self,
        host='localhost',
        port=21050,
        database='default',
        timeout=45,
        use_ssl=False,
        ca_cert=None,
        user=None,
        password=None,
        auth_mechanism='NOSASL',
        kerberos_service_name='impala',
        pool_size=8,
        hdfs_client=None,
    ):
        """Create an ImpalaClient for use with Ibis.

        Parameters
        ----------
        host : str, optional
            Host name of the impalad or HiveServer2 in Hive
        port : int, optional
            Impala's HiveServer2 port
        database : str, optional
            Default database when obtaining new cursors
        timeout : int, optional
            Connection timeout in seconds when communicating with HiveServer2
        use_ssl : bool, optional
            Use SSL when connecting to HiveServer2
        ca_cert : str, optional
            Local path to 3rd party CA certificate or copy of server
            certificate for self-signed certificates. If SSL is enabled, but
            this argument is ``None``, then certificate validation is skipped.
        user : str, optional
            LDAP user to authenticate
        password : str, optional
            LDAP password to authenticate
        auth_mechanism : str, optional
            {'NOSASL' <- default, 'PLAIN', 'GSSAPI', 'LDAP'}.
            Use NOSASL for non-secured Impala connections.  Use PLAIN for
            non-secured Hive clusters.  Use LDAP for LDAP authenticated
            connections.  Use GSSAPI for Kerberos-secured clusters.
        kerberos_service_name : str, optional
            Specify particular impalad service principal.

        Examples
        --------
        >>> import ibis
        >>> import os
        >>> hdfs_host = os.environ.get('IBIS_TEST_NN_HOST', 'localhost')
        >>> hdfs_port = int(os.environ.get('IBIS_TEST_NN_PORT', 50070))
        >>> impala_host = os.environ.get('IBIS_TEST_IMPALA_HOST', 'localhost')
        >>> impala_port = int(os.environ.get('IBIS_TEST_IMPALA_PORT', 21050))
        >>> hdfs = ibis.impala.hdfs_connect(host=hdfs_host, port=hdfs_port)
        >>> hdfs  # doctest: +ELLIPSIS
        <ibis.filesystems.WebHDFS object at 0x...>
        >>> client = ibis.impala.connect(
        ...     host=impala_host,
        ...     port=impala_port,
        ...     hdfs_client=hdfs,
        ... )
        >>> client  # doctest: +ELLIPSIS
        <ibis.impala.client.ImpalaClient object at 0x...>

        Returns
        -------
        ImpalaClient
        """
        params = {
            'host': host,
            'port': port,
            'database': database,
            'timeout': timeout,
            'use_ssl': use_ssl,
            'ca_cert': ca_cert,
            'user': user,
            'password': password,
            'auth_mechanism': auth_mechanism,
            'kerberos_service_name': kerberos_service_name,
        }

        con = ImpalaConnection(pool_size=pool_size, **params)
        try:
            client = ImpalaClient(
                backend=self, con=con, hdfs_client=hdfs_client
            )
        except Exception:
            con.close()
            raise
        return client

    def register_options(self):
        ibis.config.register_option(
            'temp_db',
            '__ibis_tmp',
            'Database to use for temporary tables, views. functions, etc.',
        )
        ibis.config.register_option(
            'temp_hdfs_path',
            '/tmp/ibis',
            'HDFS path for storage of temporary data',
        )
