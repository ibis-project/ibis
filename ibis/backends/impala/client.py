from __future__ import annotations

import contextlib
import traceback
from typing import TYPE_CHECKING

import sqlalchemy as sa
import sqlglot as sg

import ibis.common.exceptions as com
import ibis.expr.schema as sch
import ibis.expr.types as ir
from ibis import util
from ibis.backends.base import Database
from ibis.backends.base.sql.compiler import DDL, DML
from ibis.backends.base.sql.ddl import AlterTable, InsertSelect
from ibis.backends.impala import ddl
from ibis.backends.impala.compat import HS2Error, impyla

if TYPE_CHECKING:
    import pandas as pd


class ImpalaDatabase(Database):
    def create_table(self, name: str, obj=None, **kwargs) -> ir.Table:
        """Dispatch to ImpalaClient.create_table.

        See that function's docstring for more
        """
        return self.client.create_table(name, obj=obj, database=self.name, **kwargs)

    def list_udfs(self, like=None):
        return self.client.list_udfs(like=like, database=self.name)

    def list_udas(self, like=None):
        return self.client.list_udas(like=like, database=self.name)


class ImpalaConnection:
    """Database connection wrapper."""

    def __init__(self, pool_size=8, database="default", **params):
        self.params = params
        self.database = database
        self.options = {}
        self.pool = sa.pool.QueuePool(
            self._new_cursor,
            pool_size=pool_size,
            # disable invoking rollback, because any transactions in impala are
            # automatic:
            # https://impala.apache.org/docs/build/html/topics/impala_transactions.html
            reset_on_return=False,
        )

        @sa.event.listens_for(self.pool, "checkout")
        def _(dbapi_connection, *_):
            """Update `dbapi_connection` options if they don't match `self`."""
            if (options := self.options) != dbapi_connection.options:
                dbapi_connection.options = options.copy()
                dbapi_connection.set_options()

    def set_options(self, options):
        self.options.update(options)

    def close(self):
        """Close all idle Impyla connections."""
        self.pool.dispose()

    def disable_codegen(self, disabled=True):
        self.options["DISABLE_CODEGEN"] = str(int(disabled))

    def execute(self, query):
        if isinstance(query, (DDL, DML)):
            query = query.compile()

        util.log(query)

        cursor = self.pool.connect()
        try:
            cursor.execute(query)
        except Exception:
            cursor.close()
            util.log(f"Exception caused by {query}: {traceback.format_exc()}")
            raise

        return cursor

    def fetchall(self, query):
        cur = self.execute(query)
        try:
            results = cur.fetchall()
        finally:
            cur.close()
        return results

    def _new_cursor(self):
        params = self.params.copy()
        con = impyla.connect(database=self.database, **params)

        # make sure the connection works
        cursor = con.cursor(user=params.get("user"), convert_types=True)
        cursor.ping()

        wrapper = ImpalaCursor(cursor, con, self.database, self.options.copy())
        wrapper.set_options()
        return wrapper

    def release(self, cur):  # pragma: no cover
        pass


class ImpalaCursor:
    def __init__(self, cursor, impyla_con, database, options):
        self._cursor = cursor
        self.impyla_con = impyla_con
        self.database = database
        self.options = options

    def __del__(self):
        with contextlib.suppress(Exception):
            self.close()

    def close(self):
        try:
            self._cursor.close()
        except HS2Error as e:
            # connection was closed elsewhere
            already_closed_messages = [
                "invalid query handle",
                "invalid session",
            ]
            for message in already_closed_messages:
                if message in e.args[0].lower():
                    break
            else:
                raise

    def set_options(self):
        for k, v in self.options.items():
            query = f"SET {k} = {v!r}"
            self._cursor.execute(query)

    @property
    def description(self):
        return self._cursor.description

    def release(self):
        pass

    def execute(self, stmt):
        self._cursor.execute_async(stmt)
        self._wait_synchronous()

    def _wait_synchronous(self):
        # Wait to finish, but cancel if KeyboardInterrupt
        try:
            self._cursor._wait_to_finish()
        except KeyboardInterrupt:
            util.log("Canceling query")
            self.cancel()
            raise

    def is_finished(self):
        return not self.is_executing()

    def is_executing(self):
        return self._cursor.is_executing()

    def cancel(self):
        self._cursor.cancel_operation()

    def fetchone(self):
        return self._cursor.fetchone()

    def fetchall(self, columnar=False):
        if columnar:
            return self._cursor.fetchcolumnar()
        else:
            return self._cursor.fetchall()


class ImpalaTable(ir.Table):
    """A physical table in the Impala-Hive metastore."""

    @property
    def _qualified_name(self) -> str:
        op = self.op()
        return sg.table(op.name, db=op.namespace).sql(dialect="hive")

    @property
    def _unqualified_name(self) -> str:
        return self.op().name

    @property
    def _client(self):
        return self.op().source

    @property
    def _database(self) -> str:
        return self.op().namespace

    def compute_stats(self, incremental=False):
        """Invoke Impala COMPUTE STATS command on the table."""
        return self._client.compute_stats(self._qualified_name, incremental=incremental)

    def invalidate_metadata(self):
        self._client.invalidate_metadata(self._qualified_name)

    def refresh(self):
        self._client.refresh(self._qualified_name)

    def metadata(self):
        """Return results of `DESCRIBE FORMATTED` statement."""
        return self._client.describe_formatted(self._qualified_name)

    describe_formatted = metadata

    def files(self):
        """Return results of SHOW FILES statement."""
        return self._client.show_files(self._qualified_name)

    def drop(self):
        """Drop the table from the database."""
        self._client.drop_table_or_view(self._qualified_name)

    def truncate(self):
        self._client.truncate_table(self._qualified_name)

    def insert(
        self,
        obj=None,
        overwrite=False,
        partition=None,
        values=None,
        validate=True,
    ):
        """Insert into an Impala table.

        Parameters
        ----------
        obj
            Table expression or DataFrame
        overwrite
            If True, will replace existing contents of table
        partition
            For partitioned tables, indicate the partition that's being
            inserted into, either with an ordered list of partition keys or a
            dict of partition field name to value. For example for the
            partition (year=2007, month=7), this can be either (2007, 7) or
            {'year': 2007, 'month': 7}.
        values
            Unsupported and unused
        validate
            If True, do more rigorous validation that schema of table being
            inserted is compatible with the existing table

        Examples
        --------
        Append to an existing table

        >>> t.insert(table_expr)  # quartodoc: +SKIP # doctest: +SKIP

        Completely overwrite contents

        >>> t.insert(table_expr, overwrite=True)  # quartodoc: +SKIP # doctest: +SKIP
        """
        if values is not None:
            raise NotImplementedError

        with self._client._setup_insert(obj) as expr:
            if validate:
                existing_schema = self.schema()
                insert_schema = expr.schema()
                if not insert_schema.equals(existing_schema):
                    _validate_compatible(insert_schema, existing_schema)

            if partition is not None:
                partition_schema = self.partition_schema()
                partition_schema_names = frozenset(partition_schema.names)
                expr = expr.select(
                    [
                        column
                        for column in expr.columns
                        if column not in partition_schema_names
                    ]
                )
            else:
                partition_schema = None

            ast = self._client.compiler.to_ast(expr)
            select = ast.queries[0]
            statement = InsertSelect(
                self._qualified_name,
                select,
                partition=partition,
                partition_schema=partition_schema,
                overwrite=overwrite,
            )
            self._client._safe_exec_sql(statement.compile())
            return self

    def load_data(self, path, overwrite=False, partition=None):
        """Load data into an Impala table.

        Parameters
        ----------
        path
            Data to load
        overwrite
            Overwrite the existing data in the entire table or indicated
            partition
        partition
            If specified, the partition must already exist
        """
        if partition is not None:
            partition_schema = self.partition_schema()
        else:
            partition_schema = None

        stmt = ddl.LoadData(
            self._qualified_name,
            path,
            partition=partition,
            partition_schema=partition_schema,
            overwrite=overwrite,
        )

        self._client._safe_exec_sql(stmt.compile())
        return self

    @property
    def name(self) -> str:
        return self.op().name

    @property
    def is_partitioned(self):
        """True if the table is partitioned."""
        return self.metadata().is_partitioned

    def partition_schema(self):
        """Return the schema for the partition columns."""
        schema = self.schema()
        result = self.partitions()

        partition_fields = []
        for col in result.columns:
            if col not in schema:
                break
            partition_fields.append((col, schema[col]))

        return sch.Schema(dict(partition_fields))

    def add_partition(self, spec, location=None):
        """Add a new table partition.

        This API creates any necessary new directories in HDFS.

        Partition parameters can be set in a single DDL statement or you can
        use `alter_partition` to set them after the fact.
        """
        part_schema = self.partition_schema()
        stmt = ddl.AddPartition(
            self._qualified_name, spec, part_schema, location=location
        )
        self._client._safe_exec_sql(stmt)
        return self

    def alter(
        self,
        location=None,
        format=None,
        tbl_properties=None,
        serde_properties=None,
    ):
        """Change settings and parameters of the table.

        Parameters
        ----------
        location
            For partitioned tables, you may want the alter_partition function
        format
            Table format
        tbl_properties
            Table properties
        serde_properties
            Serialization/deserialization properties
        """

        def _run_ddl(**kwds):
            stmt = AlterTable(self._qualified_name, **kwds)
            self._client._safe_exec_sql(stmt)
            return self

        return self._alter_table_helper(
            _run_ddl,
            location=location,
            format=format,
            tbl_properties=tbl_properties,
            serde_properties=serde_properties,
        )

    def set_external(self, is_external=True):
        """Toggle the `EXTERNAL` table property."""
        self.alter(tbl_properties={"EXTERNAL": is_external})

    def alter_partition(
        self,
        spec,
        location=None,
        format=None,
        tbl_properties=None,
        serde_properties=None,
    ):
        """Change settings and parameters of an existing partition.

        Parameters
        ----------
        spec
            The partition keys for the partition being modified
        location
            Location of the partition
        format
            Table format
        tbl_properties
            Table properties
        serde_properties
            Serialization/deserialization properties
        """
        part_schema = self.partition_schema()

        def _run_ddl(**kwds):
            stmt = ddl.AlterPartition(self._qualified_name, spec, part_schema, **kwds)
            self._client._safe_exec_sql(stmt)
            return self

        return self._alter_table_helper(
            _run_ddl,
            location=location,
            format=format,
            tbl_properties=tbl_properties,
            serde_properties=serde_properties,
        )

    def _alter_table_helper(self, f, **alterations):
        results = []
        for k, v in alterations.items():
            if v is None:
                continue
            result = f(**{k: v})
            results.append(result)
        return results

    def drop_partition(self, spec):
        """Drop an existing table partition."""
        part_schema = self.partition_schema()
        stmt = ddl.DropPartition(self._qualified_name, spec, part_schema)
        self._client._safe_exec_sql(stmt)
        return self

    def partitions(self):
        """Return information about the table's partitions.

        Raises an exception if the table is not partitioned.
        """
        return self._client.list_partitions(self._qualified_name)

    def stats(self) -> pd.DataFrame:
        """Return results of `SHOW TABLE STATS`.

        If not partitioned, contains only one row.

        Returns
        -------
        DataFrame
            Table statistics
        """
        return self._client.table_stats(self._qualified_name)

    def column_stats(self) -> pd.DataFrame:
        """Return results of `SHOW COLUMN STATS`.

        Returns
        -------
        DataFrame
            Column statistics
        """
        return self._client.column_stats(self._qualified_name)


# ----------------------------------------------------------------------
# ORM-ish usability layer


class ScalarFunction:
    def drop(self):
        pass


class AggregateFunction:
    def drop(self):
        pass


def _validate_compatible(from_schema, to_schema):
    if set(from_schema.names) != set(to_schema.names):
        raise com.IbisInputError("Schemas have different names")

    for name in from_schema:
        lt = from_schema[name]
        rt = to_schema[name]
        if not lt.castable(rt):
            raise com.IbisInputError(f"Cannot safely cast {lt!r} to {rt!r}")
