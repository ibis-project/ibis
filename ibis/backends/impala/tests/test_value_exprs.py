from __future__ import annotations

import pandas as pd
import pytest
from pytest import param

import ibis
from ibis import literal as L
from ibis.backends.impala.compiler import ImpalaCompiler
from ibis.backends.impala.tests.conftest import translate


@pytest.fixture(scope="module")
def table(mockcon):
    return mockcon.table("alltypes")


@pytest.mark.parametrize(
    "value",
    [
        param("simple", id="simple"),
        param("I can't", id="embedded_single_quote"),
        param('An "escape"', id="embedded_double_quote"),
        param(5, id="int"),
        param(1.5, id="float"),
        param(True, id="true"),
        param(False, id="false"),
    ],
)
def test_literals(value, snapshot):
    expr = L(value)
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


def test_column_ref_table_aliases(snapshot):
    context = ImpalaCompiler.make_context()

    table1 = ibis.table([("key1", "string"), ("value1", "double")])

    table2 = ibis.table([("key2", "string"), ("value and2", "double")])

    context.set_ref(table1.op(), "t0")
    context.set_ref(table2.op(), "t1")

    expr = table1["value1"] - table2["value and2"]

    result = translate(expr, context=context)
    snapshot.assert_match(result, "out.sql")


def test_column_ref_quoting():
    schema = [("has a space", "double")]
    table = ibis.table(schema)
    translate(table["has a space"], named="`has a space`")


def test_identifier_quoting():
    schema = [("date", "double"), ("table", "string")]
    table = ibis.table(schema)
    translate(table["date"], named="`date`")
    translate(table["table"], named="`table`")


@pytest.mark.parametrize(
    "expr_fn",
    [
        lambda t: t.g.cast("double").name("g_dub"),
        lambda t: t.g.name("has a space"),
        lambda t: ((t.a - t.b) * t.a).name("expr"),
    ],
    ids=["cast", "spaces", "compound_expr"],
)
def test_named_expressions(table, expr_fn, snapshot):
    expr = expr_fn(table)
    result = translate(expr, named=True)
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize(
    "expr_fn",
    [
        param(lambda t: t.a + t.b, id="add"),
        param(lambda t: t.a - t.b, id="sub"),
        param(lambda t: t.a * t.b, id="mul"),
        param(lambda t: t.a / t.b, id="div"),
        param(lambda t: t.a**t.b, id="pow"),
        param(lambda t: t.a < t.b, id="lt"),
        param(lambda t: t.a <= t.b, id="le"),
        param(lambda t: t.a > t.b, id="gt"),
        param(lambda t: t.a >= t.b, id="ge"),
        param(lambda t: t.a == t.b, id="eq"),
        param(lambda t: t.a != t.b, id="ne"),
        param(lambda t: t.h & (t.a > 0), id="and"),
        param(lambda t: t.h | (t.a > 0), id="or"),
        param(lambda t: t.h ^ (t.a > 0), id="xor"),
    ],
)
def test_binary_infix_operators(table, expr_fn, snapshot):
    expr = expr_fn(table)
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize(
    "expr_fn",
    [
        lambda t: (t.a + t.b) + t.c,
        lambda t: t.a.log() + t.c,
        lambda t: t.b + (-(t.a + t.c)),
    ],
    ids=["parens_left", "function_call", "negation"],
)
def test_binary_infix_parenthesization(table, expr_fn, snapshot):
    expr = expr_fn(table)
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


def test_between(table, snapshot):
    expr = table.f.between(0, 1)
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize(
    "expr_fn",
    [
        lambda t: t["g"].isnull(),
        lambda t: t["a"].notnull(),
        lambda t: (t["a"] + t["b"]).isnull(),
    ],
    ids=["isnull", "notnull", "compound_isnull"],
)
def test_isnull_notnull(table, expr_fn, snapshot):
    expr = expr_fn(table)
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize(
    ("column", "to_type"),
    [
        ("a", "int16"),
        ("a", "int32"),
        ("a", "int64"),
        ("a", "string"),
        ("d", "int8"),
        ("g", "double"),
        ("g", "timestamp"),
    ],
)
def test_casts(table, column, to_type, snapshot):
    expr = table[column].cast(to_type)
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


def test_misc_conditionals(table, snapshot):
    expr = table.a.nullif(0)
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize(
    "expr_fn",
    [
        lambda _: L("9.9999999").cast("decimal(38, 5)"),
        lambda t: t.f.cast("decimal(12, 2)"),
    ],
    ids=["literal", "column"],
)
def test_decimal_casts(table, expr_fn, snapshot):
    expr = expr_fn(table)
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize("colname", ["a", "f", "h"])
def test_negate(table, colname, snapshot):
    result = translate(-table[colname])
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize(
    "field",
    ["year", "month", "day", "hour", "minute", "second", "microsecond", "millisecond"],
)
def test_timestamp_extract_field(table, field, snapshot):
    expr = getattr(table.i, field)()
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


def test_sql_extract(table, snapshot):
    # integration with SQL translation
    expr = table[
        table.i.year().name("year"),
        table.i.month().name("month"),
        table.i.day().name("day"),
    ]

    result = ImpalaCompiler.to_sql(expr)
    snapshot.assert_match(result, "out.sql")


def test_timestamp_now(snapshot):
    expr = ibis.now()
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize(
    "unit",
    ["years", "months", "weeks", "days", "hours", "minutes", "seconds"],
)
def test_timestamp_deltas(table, unit, snapshot):
    K = 5

    offset = ibis.interval(**{unit: K})

    add_expr = table.i + offset
    result = translate(add_expr)
    snapshot.assert_match(result, "out1.sql")

    sub_expr = table.i - offset
    result = translate(sub_expr)
    snapshot.assert_match(result, "out2.sql")


@pytest.mark.parametrize(
    "expr_fn",
    [
        lambda v: L(pd.Timestamp(v)),
        lambda v: L(pd.Timestamp(v).to_pydatetime()),
        lambda v: ibis.timestamp(v),
    ],
    ids=["pd_timestamp", "pydatetime", "timestamp_function"],
)
def test_timestamp_literals(expr_fn, snapshot):
    expr = expr_fn("2015-01-01 12:34:56")
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize("method_name", ["index", "full_name"])
def test_timestamp_day_of_week(method_name, snapshot):
    ts = ibis.timestamp("2015-09-01T01:00:23")
    expr = getattr(ts.day_of_week, method_name)()
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize(
    "expr_fn",
    [
        lambda col: col.to_timestamp(),
        lambda col: col.to_timestamp("ms"),
        lambda col: col.to_timestamp("us"),
    ],
    ids=["default", "ms", "us"],
)
def test_timestamp_from_integer(table, expr_fn, snapshot):
    expr = expr_fn(table.c)
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


def test_correlated_predicate_subquery(table, snapshot):
    t0 = table
    t1 = t0.view()

    expr = t0.g == t1.g

    ctx = ImpalaCompiler.make_context()
    ctx.make_alias(t0.op())

    # Grab alias from parent context
    subctx = ctx.subcontext()
    subctx.make_alias(t1.op())
    subctx.make_alias(t0.op())

    result = translate(expr, context=subctx)
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize(
    "expr_fn",
    [
        param(lambda b: b.any(), id="any"),
        param(lambda b: -b.any(), id="not_any"),
        param(lambda b: b.all(), id="all"),
        param(lambda b: -b.all(), id="not_all"),
    ],
)
def test_any_all(table, expr_fn, snapshot):
    expr = expr_fn(table.f == 0)
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")
