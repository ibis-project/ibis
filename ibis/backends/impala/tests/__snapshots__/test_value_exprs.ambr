# name: test_any_all[all]
  'min(`f` = 0)'
# ---
# name: test_any_all[any]
  'max(`f` = 0)'
# ---
# name: test_any_all[not_all]
  'min(`f` = 0) = FALSE'
# ---
# name: test_any_all[not_any]
  'max(`f` = 0) = FALSE'
# ---
# name: test_between
  '`f` BETWEEN 0 AND 1'
# ---
# name: test_binary_infix_operators[add]
  '`a` + `b`'
# ---
# name: test_binary_infix_operators[and]
  '`h` AND (`a` > 0)'
# ---
# name: test_binary_infix_operators[div]
  '`a` / `b`'
# ---
# name: test_binary_infix_operators[eq]
  '`a` = `b`'
# ---
# name: test_binary_infix_operators[ge]
  '`a` >= `b`'
# ---
# name: test_binary_infix_operators[gt]
  '`a` > `b`'
# ---
# name: test_binary_infix_operators[le]
  '`a` <= `b`'
# ---
# name: test_binary_infix_operators[lt]
  '`a` < `b`'
# ---
# name: test_binary_infix_operators[mul]
  '`a` * `b`'
# ---
# name: test_binary_infix_operators[ne]
  '`a` != `b`'
# ---
# name: test_binary_infix_operators[or]
  '`h` OR (`a` > 0)'
# ---
# name: test_binary_infix_operators[pow]
  'pow(`a`, `b`)'
# ---
# name: test_binary_infix_operators[sub]
  '`a` - `b`'
# ---
# name: test_binary_infix_operators[xor]
  '(`h` OR (`a` > 0)) AND NOT (`h` AND (`a` > 0))'
# ---
# name: test_binary_infix_parenthesization[function_call]
  'ln(`a`) + `c`'
# ---
# name: test_binary_infix_parenthesization[negation]
  '`b` + (-(`a` + `c`))'
# ---
# name: test_binary_infix_parenthesization[parens_left]
  '(`a` + `b`) + `c`'
# ---
# name: test_casts[a-int16]
  'CAST(`a` AS smallint)'
# ---
# name: test_casts[a-int32]
  'CAST(`a` AS int)'
# ---
# name: test_casts[a-int64]
  'CAST(`a` AS bigint)'
# ---
# name: test_casts[a-string]
  'CAST(`a` AS string)'
# ---
# name: test_casts[d-int8]
  'CAST(`d` AS tinyint)'
# ---
# name: test_casts[g-double]
  'CAST(`g` AS double)'
# ---
# name: test_casts[g-timestamp]
  'CAST(`g` AS timestamp)'
# ---
# name: test_column_ref_table_aliases
  't0.`value1` - t1.`value and2`'
# ---
# name: test_correlated_predicate_subquery
  't0.`g` = t1.`g`'
# ---
# name: test_decimal_builtins[precision]
  'precision(`l_extendedprice`)'
# ---
# name: test_decimal_builtins[scale]
  'scale(`l_extendedprice`)'
# ---
# name: test_decimal_casts[column]
  'CAST(`f` AS decimal(12, 2))'
# ---
# name: test_decimal_casts[literal]
  "CAST('9.9999999' AS decimal(38, 5))"
# ---
# name: test_isnull_notnull[compound_isnull]
  '`a` + `b` IS NULL'
# ---
# name: test_isnull_notnull[isnull]
  '`g` IS NULL'
# ---
# name: test_isnull_notnull[notnull]
  '`a` IS NOT NULL'
# ---
# name: test_literals[embedded_double_quote]
  '\'An "escape"\''
# ---
# name: test_literals[embedded_single_quote]
  "'I can\\'t'"
# ---
# name: test_literals[false]
  'FALSE'
# ---
# name: test_literals[float]
  '1.5'
# ---
# name: test_literals[int]
  '5'
# ---
# name: test_literals[simple]
  "'simple'"
# ---
# name: test_literals[true]
  'TRUE'
# ---
# name: test_misc_conditionals
  'nullif(`a`, 0)'
# ---
# name: test_named_expressions[cast]
  'CAST(`g` AS double) AS `g_dub`'
# ---
# name: test_named_expressions[compound_expr]
  '(`a` - `b`) * `a` AS `expr`'
# ---
# name: test_named_expressions[spaces]
  '`g` AS `has a space`'
# ---
# name: test_negate[a]
  '-`a`'
# ---
# name: test_negate[f]
  '-`f`'
# ---
# name: test_negate[h]
  'NOT `h`'
# ---
# name: test_sql_extract
  '''
  SELECT extract(`i`, 'year') AS `year`, extract(`i`, 'month') AS `month`,
         extract(`i`, 'day') AS `day`
  FROM alltypes
  '''
# ---
# name: test_timestamp_day_of_week[full_name]
  "dayname('2015-09-01T01:00:23')"
# ---
# name: test_timestamp_day_of_week[index]
  "pmod(dayofweek('2015-09-01T01:00:23') - 2, 7)"
# ---
# name: test_timestamp_deltas[days]
  'date_add(cast(`i` as timestamp), INTERVAL 5 DAY)'
# ---
# name: test_timestamp_deltas[days].1
  'date_sub(cast(`i` as timestamp), INTERVAL 5 DAY)'
# ---
# name: test_timestamp_deltas[hours]
  'date_add(cast(`i` as timestamp), INTERVAL 5 HOUR)'
# ---
# name: test_timestamp_deltas[hours].1
  'date_sub(cast(`i` as timestamp), INTERVAL 5 HOUR)'
# ---
# name: test_timestamp_deltas[minutes]
  'date_add(cast(`i` as timestamp), INTERVAL 5 MINUTE)'
# ---
# name: test_timestamp_deltas[minutes].1
  'date_sub(cast(`i` as timestamp), INTERVAL 5 MINUTE)'
# ---
# name: test_timestamp_deltas[months]
  'date_add(cast(`i` as timestamp), INTERVAL 5 MONTH)'
# ---
# name: test_timestamp_deltas[months].1
  'date_sub(cast(`i` as timestamp), INTERVAL 5 MONTH)'
# ---
# name: test_timestamp_deltas[seconds]
  'date_add(cast(`i` as timestamp), INTERVAL 5 SECOND)'
# ---
# name: test_timestamp_deltas[seconds].1
  'date_sub(cast(`i` as timestamp), INTERVAL 5 SECOND)'
# ---
# name: test_timestamp_deltas[weeks]
  'date_add(cast(`i` as timestamp), INTERVAL 5 WEEK)'
# ---
# name: test_timestamp_deltas[weeks].1
  'date_sub(cast(`i` as timestamp), INTERVAL 5 WEEK)'
# ---
# name: test_timestamp_deltas[years]
  'date_add(cast(`i` as timestamp), INTERVAL 5 YEAR)'
# ---
# name: test_timestamp_deltas[years].1
  'date_sub(cast(`i` as timestamp), INTERVAL 5 YEAR)'
# ---
# name: test_timestamp_extract_field[day]
  "extract(`i`, 'day')"
# ---
# name: test_timestamp_extract_field[hour]
  "extract(`i`, 'hour')"
# ---
# name: test_timestamp_extract_field[millisecond]
  "extract(`i`, 'millisecond')"
# ---
# name: test_timestamp_extract_field[minute]
  "extract(`i`, 'minute')"
# ---
# name: test_timestamp_extract_field[month]
  "extract(`i`, 'month')"
# ---
# name: test_timestamp_extract_field[second]
  "extract(`i`, 'second')"
# ---
# name: test_timestamp_extract_field[year]
  "extract(`i`, 'year')"
# ---
# name: test_timestamp_from_integer[default]
  'CAST(from_unixtime(`c`, "yyyy-MM-dd HH:mm:ss") AS timestamp)'
# ---
# name: test_timestamp_from_integer[ms]
  'CAST(from_unixtime(CAST(floor(`c` / 1000) AS int), "yyyy-MM-dd HH:mm:ss") AS timestamp)'
# ---
# name: test_timestamp_from_integer[us]
  'CAST(from_unixtime(CAST(floor(`c` / 1000000) AS int), "yyyy-MM-dd HH:mm:ss") AS timestamp)'
# ---
# name: test_timestamp_literals[pd_timestamp]
  "'2015-01-01T12:34:56'"
# ---
# name: test_timestamp_literals[pydatetime]
  "'2015-01-01T12:34:56'"
# ---
# name: test_timestamp_literals[timestamp_function]
  "'2015-01-01T12:34:56'"
# ---
# name: test_timestamp_now
  'now()'
# ---
